server/sonar-server/src/main/java/org/sonar/server/computation/issue/commonrule/AbstractCoverageRule.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.issue.commonrule;
+
+import com.google.common.base.Optional;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.measure.Measure;
+import org.sonar.server.computation.measure.MeasureRepository;
+import org.sonar.server.computation.metric.Metric;
+import org.sonar.server.computation.qualityprofile.ActiveRule;
+import org.sonar.server.computation.qualityprofile.ActiveRulesHolder;
+
+public abstract class AbstractCoverageRule extends CommonRule {
+
+  private final MeasureRepository measureRepository;
+  private final Metric coverageMetric;
+  private final Metric uncoveredMetric;
+  private final Metric toCoverMetric;
+
+  public AbstractCoverageRule(ActiveRulesHolder activeRulesHolder, String ruleKey,
+    MeasureRepository measureRepository,
+    Metric coverageMetric, Metric uncoveredMetric, Metric toCoverMetric) {
+    super(activeRulesHolder, ruleKey);
+    this.measureRepository = measureRepository;
+    this.coverageMetric = coverageMetric;
+    this.uncoveredMetric = uncoveredMetric;
+    this.toCoverMetric = toCoverMetric;
+  }
+
+  @Override
+  protected CommonRuleIssue doProcessFile(Component file, ActiveRule activeRule) {
+    Optional<Measure> coverageMeasure = measureRepository.getRawMeasure(file, coverageMetric);
+    if (!file.getFileAttributes().isUnitTest() && coverageMeasure.isPresent()) {
+      double coverage = coverageMeasure.get().getDoubleValue();
+      // FIXME load threshold from activeRule
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)